File: src/Baseline.hs
{-# LANGUAGE ForeignFunctionInterface #-}
module Baseline where

import Foreign.C

foreign import ccall unsafe "nthIntegral" c_nthIntegral :: CDouble -> CInt -> CDouble

nthIntegral :: Double -> Int -> Double
nthIntegral rate n = realToFrac $ c_nthIntegral (realToFrac rate) (fromIntegral n)

File: cbits/Baseline.c
#include <math.h>
#include "Baseline.h"

double nthIntegral(double dt, int n)
{
	double integral = 0;
	int i;
	for(i = 0; i < n; i++) {
		double t = (double)i / dt;
		double s = sin(t);
		integral = integral + s * dt;
	}
	return integral;
}
